---
title: 3FED9882-642E-41CF-933C-1E4ACB0E1BC9
tags:
  - js
  - node
---

# もしかしてaa functionって例外処理必要ないんじゃないの？？？

ちょっと釣りタイトルっぽくなってしまいましたが以下の前提条件での検証を行ってみました。

- `return new Promise`の代わりに`async function`を使う

検証はnodejsです。

## 前提条件

- `return new Promise`はPromiseオブジェクトを返す
- `async function`はPromiseオブジェクトを返す

まず以下のコードを御覧ください。

```js
const fs = require('fs');
module.exports = function readFile(_path) {
  return new Promise((resolve, reject) => {
    fs.readFile(_path, 'utf8', (err, data) => {
      if (err) {
        reject(err);
      }
      resolve(data);
    });
  });
};

```

`fs.readFile`をPromise化しただけの単純なものです。これを`async function`に変換することは不可能です。
`async function`は`await`でしか処理を待つことができないのでコールバックによる非同期処理を行っている`fs.readFile`を`async function`にすることは不可能です。  
(もちろんawaitする為のPromiseを別に書けば可能です)

しかし`fs.readFile`と同等の機能を持つ`fs.readFileSync`を使えば`async function`を使いPromise化することができるのではないでしょうか。

```js
const fs = require('fs');
module.exports = async function readFile(_path) {
  return fs.readFileSync(_path, 'utf8');
};

```

node.jsのネイティブクラスであるfsは`~Sync`というMethodが多くあります。上記の例では`readFile`が非同期Methodであるのに対し`readFileSync`は同期Methodです。

これらの説明をしているサイトでは「`readFileSync`ではなく`readFile`を使おうね！」という記述がよく見られます。理由は主に以下の通り

- `~Sync`系は処理をブロックするのでノンブロッキングI/O恩恵を受けられない
- エラーハンドリング忘れがち
- try/catch書くのめんどくさい
- try/catchどこに書くのか問題

このうちエラーハンドリングの問題は`async function`の場合必要ありません。なのでSync系のMethodを使うことへの抵抗が薄れます。

## async functionの基本的仕様

- async functionはPromiseを返す
- returnした値はResolveされる
- throwされたらRejectされる
- awaitで処理を待つ


|          |   new Promise   | async function |
|----------|:---------------:|---------------:|
| 戻り値   | resolve(戻り値) |  return 戻り値 |
| 例外     |  reject(error)  |           none |

## メリデメ

### `new Promise`のメリット

- callbackを待ってからresolveできる

### `new Promise`のデメリット

- resolveをreturnするの忘れる問題


#### resolveをreturnするの忘れる問題

処理の途中でresolveした時に`return resolve(foo)`というように`return`を忘れると処理が続いてしまう。(ただし防止策としてeslintで警告を出せる)


### `async function`のメリット

- return すると自動でresolveする

#### return すると自動でresolveする

そもそも`async function`の仕様なのですが、`return foo`は`return resolve(foo)`と同義です。

### `async function`のデメリット

- callback待てない問題
- awaitにはPromiseが必要